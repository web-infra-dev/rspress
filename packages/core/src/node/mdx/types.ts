import type { FrontMatterMeta, UserConfig } from '@rspress/shared';

import type { PluginDriver } from '../PluginDriver';
import type { RouteService } from '../route/RouteService';
import type { TocItem } from './remarkPlugins/toc';

export interface MdxLoaderOptions {
  config: UserConfig;
  docDirectory: string;
  checkDeadLinks: boolean;
  routeService: RouteService;
  pluginDriver: PluginDriver;
}

export interface PageMeta {
  toc: TocItem[];
  title: string;
  headingTitle: string;
  frontmatter?: FrontMatterMeta;
}
