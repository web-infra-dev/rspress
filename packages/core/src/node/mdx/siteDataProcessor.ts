import type { Header } from '@rspress/shared';
import rehypeStringify from 'rehype-stringify';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import { unified } from 'unified';
import type { RouteService } from '../route/RouteService';
import { rehypeHeaderAnchor } from './rehypePlugins/headerAnchor';
import { remarkPluginNormalizeLink } from './remarkPlugins/normalizeLink';
import { remarkPluginToc } from './remarkPlugins/toc';

async function compile(options: {
  source: string;
  root: string;
  routeService: RouteService;
}): Promise<{
  toc: Header[];
  title: string;
  html: string;
}> {
  const { source, root, routeService } = options;

  const processor = unified()
    .use(remarkParse)
    .use(remarkPluginToc)
    .use(remarkPluginNormalizeLink, {
      cleanUrls: true,
      root,
      routeService,
    })
    .use(remarkRehype)
    .use(rehypeHeaderAnchor)
    .use(rehypeStringify as any);

  processor.data('pageMeta' as any, { toc: [], title: '' });
  const html = (await processor.process(source)).toString();
  const compilationMeta = processor.data('pageMeta' as any) as {
    toc: Header[];
    title: string;
  };

  return {
    ...compilationMeta,
    html,
  };
}

export { compile };
