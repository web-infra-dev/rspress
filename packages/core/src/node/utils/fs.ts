import fs from 'node:fs/promises';

export async function pathExists(path: string): Promise<boolean> {
  try {
    await fs.access(path);
    return true;
  } catch {
    return false;
  }
}

export async function readJson(path: string): Promise<any> {
  const raw = await fs.readFile(path, 'utf8');
  return JSON.parse(raw);
}
