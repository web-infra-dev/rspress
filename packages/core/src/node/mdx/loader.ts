import path from 'path';
import type { Rspack } from '@rsbuild/core';
import { createProcessor } from '@mdx-js/mdx';
import type { Header, UserConfig } from '@rspress/shared';
import { logger } from '@rspress/shared/logger';
import { loadFrontMatter } from '@rspress/shared/node-utils';
import type { RouteService } from '../route/RouteService';
import { normalizePath, escapeMarkdownHeadingIds } from '../utils';
import { PluginDriver } from '../PluginDriver';
import { createMDXOptions } from './options';
import { TocItem } from './remarkPlugins/toc';
import { checkLinks } from './remarkPlugins/checkDeadLink';

interface LoaderOptions {
  config: UserConfig;
  docDirectory: string;
  checkDeadLinks: boolean;
  enableMdxRs: boolean;
  routeService: RouteService;
  pluginDriver: PluginDriver;
}

export interface PageMeta {
  toc: TocItem[];
  title: string;
  frontmatter?: Record<string, any>;
}

export default async function mdxLoader(
  context: Rspack.LoaderContext<LoaderOptions>,
  source: string,
  callback: Rspack.LoaderContext['callback'],
) {
  const options = context.getOptions();
  const filepath = context.resourcePath;
  context.cacheable(true);

  let pageMeta = {
    title: '',
    toc: [],
  } as PageMeta;

  const {
    config,
    docDirectory,
    checkDeadLinks,
    routeService,
    enableMdxRs,
    pluginDriver,
  } = options;

  const { frontmatter, content } = loadFrontMatter(
    source,
    filepath,
    docDirectory,
    true,
  );

  // preprocessor
  const preprocessedContent = escapeMarkdownHeadingIds(content);
  const isHomePage = frontmatter.pageType === 'home';

  try {
    let compileResult: string;
    if (!enableMdxRs) {
      const mdxOptions = await createMDXOptions(
        docDirectory,
        config,
        checkDeadLinks,
        routeService,
        filepath,
        pluginDriver,
      );
      const compiler = createProcessor(mdxOptions);

      compiler.data('pageMeta', {
        toc: [],
        title: '',
      });
      const vFile = await compiler.process({
        value: preprocessedContent,
        path: filepath,
      });

      compileResult = String(vFile);
      pageMeta = {
        ...(compiler.data('pageMeta') as {
          toc: Header[];
          title: string;
        }),
        frontmatter,
      } as PageMeta;
    } else {
      const { compile } = require('@rspress/mdx-rs');

      // TODO: Cannot get correct toc from mdx which has internal components
      const { toc, links, title, code } = await compile({
        value: preprocessedContent,
        filepath,
        root: docDirectory,
        development: process.env.NODE_ENV !== 'production',
      });

      compileResult = code;
      pageMeta = {
        toc,
        title,
        frontmatter,
      };
      // We should check dead links in mdx-rs mode
      if (checkDeadLinks) {
        checkLinks(links, filepath, docDirectory, routeService);
      }
    }

    // encode filename to be compatible with Windows
    const result = `${compileResult}
MDXContent.__RSPRESS_PAGE_META = {};
MDXContent.__RSPRESS_PAGE_META["${encodeURIComponent(
      normalizePath(path.relative(docDirectory, filepath)),
    )}"] = ${JSON.stringify(pageMeta)};
${
  // Fix the home page won't update when the frontmatter is changed
  isHomePage ? `export const frontmatter = ${JSON.stringify(frontmatter)};` : ''
}

`;
    callback(null, result);
  } catch (e) {
    logger.error(`MDX compile error: ${e.message} in ${filepath}`);
    callback({ message: e.message, name: `${filepath} compile error` });
  }
}
