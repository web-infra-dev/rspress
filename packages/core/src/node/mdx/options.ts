import path from 'path';
import remarkGFM from 'remark-gfm';
import rehypePluginExternalLinks from 'rehype-external-links';
import { PluggableList } from 'unified';
import { Options } from '@mdx-js/loader';
import { UserConfig } from '@rspress/shared';
import { remarkPluginToc } from './remarkPlugins/toc';
import { remarkBuiltin } from './remarkPlugins/builtin';
import { rehypePluginCodeMeta } from './rehypePlugins/codeMeta';
import { rehypeHeaderAnchor } from './rehypePlugins/headerAnchor';
import { remarkCheckDeadLinks } from './remarkPlugins/checkDeadLink';
import { remarkPluginNormalizeLink } from './remarkPlugins/normalizeLink';

import type { PluginDriver } from '../PluginDriver';
import type { RouteService } from '../route/RouteService';

export async function createMDXOptions(
  docDirectory: string,
  config: UserConfig,
  checkDeadLinks: boolean,
  routeService: RouteService,
  filepath: string,
  pluginDriver: PluginDriver,
): Promise<Options> {
  const cleanUrls = Boolean(config?.route?.cleanUrls);
  const {
    remarkPlugins: remarkPluginsFromConfig = [],
    rehypePlugins: rehypePluginsFromConfig = [],
    globalComponents: globalComponentsFromConfig = [],
  } = config?.markdown || {};
  const rspressPlugins = pluginDriver.getPlugins();
  const remarkPluginsFromPlugins = rspressPlugins.flatMap(
    plugin => plugin.markdown?.remarkPlugins || [],
  ) as PluggableList;
  const rehypePluginsFromPlugins = rspressPlugins.flatMap(
    plugin => plugin.markdown?.rehypePlugins || [],
  ) as PluggableList;
  const globalComponents = [
    ...rspressPlugins.flatMap(
      plugin => plugin.markdown?.globalComponents || [],
    ),
    ...globalComponentsFromConfig,
  ];
  const defaultLang = config?.lang || '';
  return {
    providerImportSource: '@mdx-js/react',
    format: path.extname(filepath).slice(1) as 'mdx' | 'md',
    remarkPlugins: [
      remarkGFM,
      remarkPluginToc,
      [
        remarkPluginNormalizeLink,
        {
          base: config?.base || '',
          cleanUrls,
          defaultLang,
          root: docDirectory,
        },
      ],
      checkDeadLinks && [
        remarkCheckDeadLinks,
        {
          root: docDirectory,
          base: config?.base || '',
          routeService,
        },
      ],
      globalComponents.length && [
        remarkBuiltin,
        {
          globalComponents,
        },
      ],
      ...remarkPluginsFromConfig,
      ...remarkPluginsFromPlugins,
    ].filter(Boolean) as PluggableList,
    rehypePlugins: [
      rehypeHeaderAnchor,
      rehypePluginCodeMeta,
      [
        rehypePluginExternalLinks,
        {
          target: '_blank',
          rel: 'noopener noreferrer',
        },
      ],
      ...rehypePluginsFromConfig,
      ...rehypePluginsFromPlugins,
    ],
  };
}
