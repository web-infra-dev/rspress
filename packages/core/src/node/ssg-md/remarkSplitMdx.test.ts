import fs from 'node:fs/promises';
import { compile } from '@mdx-js/mdx';
import { describe, expect, it } from 'vitest';
import { remarkSplitMdx } from './remarkSplitMdx';

/**
 * Process MDX with custom remark plugin
 */
export async function processMdx(source: string): Promise<string> {
  // Compile MDX with our custom remark plugin
  const result = await compile(source, {
    remarkPlugins: [remarkSplitMdx],
    jsx: true,
  });

  // Get the compiled code - this is the original MDX output
  const code = String(result);

  return code;
}

describe('mdx-to-md-loader', () => {
  it('should transform MDX with import and JSX component to template literal', async () => {
    const input = `# title

import Foo from '@components'

<Foo />`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Foo from '@components';
      function _createMdxContent(props) {
        return <><>{"# title\\n"}</>{"\\n"}{"\\n"}<Foo /></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle MDX with props', async () => {
    const input = `# Hello World

import Button from '@components/Button'

<Button type="primary" disabled={true}>Click me</Button>`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Button from '@components/Button';
      function _createMdxContent(props) {
        return <><>{"# Hello World\\n"}</>{"\\n"}{"\\n"}<Button type="primary" disabled={true}>{"Click me"}</Button></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle multiple components', async () => {
    const input = `# Documentation

import Foo from '@components/Foo'
import Bar from '@components/Bar'

Some text here.

<Foo />

More content.

<Bar prop="value" />`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Foo from '@components/Foo';
      import Bar from '@components/Bar';
      function _createMdxContent(props) {
        return <><>{"# Documentation\\n"}</>{"\\n"}{"\\n"}<>{"Some text here.\\n"}</>{"\\n"}<Foo />{"\\n"}<>{"More content.\\n"}</>{"\\n"}<Bar prop="value" /></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle codeblock', async () => {
    const input = `# Code Example
    
\`\`\`tsx
console.log('Hello, world!');
function greet(name: string) {
  return \`Hello, \${name}!\`;
}
\`\`\`
`;
    const result = await processMdx(input);
    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      function _createMdxContent(props) {
        return <><>{"# Code Example\\n"}</>{"\\n"}<>{"\`\`\`tsx\\nconsole.log('Hello, world!');\\nfunction greet(name: string) {\\n  return \`Hello, \${name}!\`;\\n}\\n\`\`\`\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle self-closing and non-self-closing components', async () => {
    const input = `
# title

import Card from '@components'

<Card>Content inside</Card>

Content outside

<Card>

\`\`\`tsx
console.log('Hello, world!');
\`\`\`
</Card>

\`\`\`tsx
console.log('Hello, world!');
\`\`\`
`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Card from '@components';
      function _createMdxContent(props) {
        const _components = {
          code: "code",
          pre: "pre",
          ...props.components
        };
        return <><>{"# title\\n"}</>{"\\n"}{"\\n"}<Card>{"Content inside"}</Card>{"\\n"}<>{"Content outside\\n"}</>{"\\n"}<Card><_components.pre><_components.code className="language-tsx">{"console.log('Hello, world!');\\n"}</_components.code></_components.pre></Card>{"\\n"}<>{"\`\`\`tsx\\nconsole.log('Hello, world!');\\n\`\`\`\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle MDX with image', async () => {
    const input = `# Image Example

import Img from '@components/Image'

import Svg from '@assets/image.svg'

Here is an image:

<Img src="/path/to/image.jpg" alt="An image" />

<img src={Svg} alt="An SVG image" />

End of content.`;

    const result = await processMdx(input);
    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Img from '@components/Image';
      import Svg from '@assets/image.svg';
      function _createMdxContent(props) {
        return <><>{"# Image Example\\n"}</>{"\\n"}{"\\n"}{"\\n"}<>{"Here is an image:\\n"}</>{"\\n"}<Img src="/path/to/image.jpg" alt="An image" />{"\\n"}<img src={Svg} alt="An SVG image" />{"\\n"}<>{"End of content.\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });
});

describe('remarkWrapMarkdown with filters', () => {
  it('should filter imports by rule - includes only @lynx', async () => {
    const input = `import { Table } from '@lynx';
import Button from 'react';

<Table />
<Button>Click</Button>`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [[['Table'], '@lynx']],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table} from '@lynx';
      import Button from 'react';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<>{"<Button>\\n  Click\\n</Button>\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should filter imports by multiple specifiers', async () => {
    const input = `import { Table, Card } from '@lynx';
import Button from 'react';

<Table />
<Card />
<Button>Click</Button>`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [
              [['Table', 'Button'], '@lynx'],
              [['Button'], 'react'],
            ],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table, Card} from '@lynx';
      import Button from 'react';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<>{"<Card />\\n"}</>{"\\n"}<Button>{"Click"}</Button></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should filter imports by source - excludes react', async () => {
    const input = `import { Table } from '@lynx';
import Button from 'react';

<Table />
<Button>Click</Button>`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            excludes: [[['Button'], 'react']],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table} from '@lynx';
      import Button from 'react';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<>{"<Button>\\n  Click\\n</Button>\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should filter imports by multiple specifiers in exclude', async () => {
    const input = `import { Table, Card } from '@lynx';

<Table />
<Card />`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            excludes: [[['Card'], '@lynx']],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table, Card} from '@lynx';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<>{"<Card />\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle import aliases correctly', async () => {
    const input = `import { Table as Tab } from '@lynx';

<Tab />`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [
              [['Table'], '@lynx'], // Match the local name (alias)
            ],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table as Tab} from '@lynx';
      function _createMdxContent(props) {
        return <><>{"<Tab />\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle combined includes and excludes - excludes takes precedence', async () => {
    const input = `import { Table, Card } from '@lynx';
import Button from 'react';

<Table />
<Card />
<Button>Click</Button>`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [[['Table', 'Card'], '@lynx']],
            excludes: [[['Card'], '@lynx']],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table, Card} from '@lynx';
      import Button from 'react';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<>{"<Card />\\n"}</>{"\\n"}<>{"<Button>\\n  Click\\n</Button>\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should support multiple include rules', async () => {
    const input = `import { Table } from '@lynx';
import { Button } from 'antd';
import Card from 'react';

<Table />
<Button />
<Card />`;

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [
              [['Table'], '@lynx'],
              [['Button'], 'antd'],
            ],
          },
        ],
      ],
      jsx: true,
    });

    const code = String(result);
    expect(code).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Table} from '@lynx';
      import {Button} from 'antd';
      import Card from 'react';
      function _createMdxContent(props) {
        return <><Table />{"\\n"}<Button />{"\\n"}<>{"<Card />\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle heading with inline JSX component', async () => {
    const input = `import Foo from '@components'

# title <Foo />

# title [link](./link)

# title [link](./link) <Foo />

Some text here.`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import Foo from '@components';
      function _createMdxContent(props) {
        return <><>{"# title "}<Foo /></>{"\\n"}<>{"# title [link](./link)\\n"}</>{"\\n"}<>{"# title [link](./link)\\n "}<Foo /></>{"\\n"}<>{"Some text here.\\n"}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should handle paragraph with multiple inline JSX components', async () => {
    const input = `import { Button, Link } from '@components'

This is a <Button>click me</Button> and <Link href="/test">link</Link> example.`;

    const result = await processMdx(input);

    expect(result).toMatchInlineSnapshot(`
      "/*@jsxRuntime automatic*/
      /*@jsxImportSource react*/
      import {Button, Link} from '@components';
      function _createMdxContent(props) {
        return <><>{"This is a "}<Button>{"click me"}</Button>{" and "}<Link href="/test">{"link"}</Link>{" example."}</></>;
      }
      export default function MDXContent(props = {}) {
        const {wrapper: MDXLayout} = props.components || ({});
        return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
      }
      "
    `);
  });

  it('should lynx', async () => {
    const fixturesTestMdxPath = new URL('./fixtures/test.mdx', import.meta.url);
    const input = await fs.readFile(fixturesTestMdxPath, 'utf-8');

    const result = await compile(input, {
      remarkPlugins: [
        [
          remarkSplitMdx,
          {
            includes: [[['Go'], '@lynx']],
          },
        ],
      ],
      jsx: true,
    });
    expect(result.toString()).toMatchSnapshot();
  });
});
