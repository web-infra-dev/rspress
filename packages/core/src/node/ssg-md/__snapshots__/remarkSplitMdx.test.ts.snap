// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkWrapMarkdown with filters > should lynx 1`] = `
"/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
import {Required, Deprecated, AndroidOnly, IOSOnly, Go, LegacyCompatTable} from '@lynx';
import {PropsTable, EventsTable, MethodsTable} from '@theme/reference';
import {InputProps, InputEvents, InputMethods} from '@theme/reference/generated/input';
function _createMdxContent(props) {
  return <>{"***\\n"}{"\\n"}{"## context: 'XElement'\\n"}{"\\n"}{"# \`<input>\`\\n"}{"\\n"}{"\\n"}{"\`<input>\` 用于创建交互式输入控件，允许用户输入和编辑单行文本。\\n"}{"\\n"}{":::info\\n"}{"\\n"}{"此功能需要客户端添加额外依赖才能被使用，集成方式请请参考 [More Elements](/guide/start/integrate-with-existing-apps).\\n"}{"\\n"}{":::\\n"}{"\\n"}{"## 使用指南\\n"}{"\\n"}{"### 基本用法\\n"}{"\\n"}{"下面是一个基本的 \`<input>\` 组件用法示例：\\n"}{"\\n"}<Go example="input" defaultFile="src/base/index.tsx" img="https://lf-lynx.tiktok-cdns.com/obj/lynx-artifacts-oss-sg/plugin/static/input_basic.PNG" highlight="{13-20}" />{"\\n"}{"### 避让键盘\\n"}{"\\n"}{"\`<input>\`不会自动的避让键盘，但可以通过监听键盘事件获取相应的高度，并以此改变\`<input>\`的位置，来进行避让：\\n"}{"\\n"}<Go example="input" defaultFile="src/auto-height/index.tsx" img="https://lf-lynx.tiktok-cdns.com/obj/lynx-artifacts-oss-sg/plugin/static/input_auto_height.PNG" highlight="{116-123}" />{"\\n"}{"## 属性\\n"}{"\\n"}{"属性名和属性值用于描述元件的行为和外观。\\n"}{"\\n"}{"### \`placeholder\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: undefined\\nplaceholder?: string\\n\`\`\`\\n"}{"\\n"}{"占位文字\\n"}{"\\n"}{"### \`confirm-type\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: undefined\\n'confirm-type'?: 'send' | 'search' | 'go' | 'done' | 'next';\\n\`\`\`\\n"}{"\\n"}{"指定输入法回车键的表现形式\\n"}{"\\n"}{"### \`maxlength\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: 140\\nmaxlength?: number;\\n\`\`\`\\n"}{"\\n"}{"输入框最大字符数量限制\\n"}{"\\n"}{"### \`readonly\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: false\\nreadonly?: boolean;\\n\`\`\`\\n"}{"\\n"}{"是否允许输入框可交互，不影响对其进行组件方法调用\\n"}{"\\n"}{"### \`show-soft-input-on-focus\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: true\\n'show-soft-input-on-focus'?: boolean;\\n\`\`\`\\n"}{"\\n"}{"聚焦时是否允许拉起系统软键盘\\n"}{"\\n"}{"### \`type\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: 'text'\\ntype?: 'text' | 'number' | 'digit' | 'password' | 'tel' | 'email';\\n\`\`\`\\n"}{"\\n"}{"指定键盘的类型\\n"}{"\\n"}{"- text\\\\tel\\\\email: 全符号\\n- number: 0-9 +-.\\n- digit: 0-9 .\\n- password: 内容变为 \\\\* 号，全符号\\n"}{"\\n"}{"::: info\\n不同的输入法可能会有不同的视觉表现\\n:::\\n"}{"\\n"}{"### \`input-filter\`\\n"}{"\\n"}{"\`\`\`ts\\n// DefaultValue: undefined\\n'input-filter'?: string;\\n\`\`\`\\n"}{"\\n"}{"指定单个字符的过滤条件，用正则表达式描述\\n"}{"\\n"}{"## 事件\\n"}{"\\n"}{"前端可以在元件上设置[事件处理器属性](../../../guide/interaction/event-handling/event-propagation.mdx#事件处理器属性)来监听元件的运行时行为。\\n"}{"\\n"}{"### \`bindfocus\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputFocusEvent {\\n  value: string;\\n}\\n bindfocus?: (e: BaseEvent<'bindfocus', InputFocusEvent>) => void;\\n\`\`\`\\n"}{"\\n"}{"输入框聚焦时的回调\\n"}{"\\n"}{"### \`bindblur\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputBlurEvent {\\n  value: string;\\n}\\nbindblur?: (e: BaseEvent<'bindblur', InputBlurEvent>) => void;\\n\`\`\`\\n"}{"\\n"}{"输入框失去焦点时的回调\\n"}{"\\n"}{"### \`bindconfirm\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputConfirmEvent {\\n  value: string;\\n}\\nbindconfirm?: (e: BaseEvent<'bindconfirm', InputConfirmEvent>) => void;\\n\`\`\`\\n"}{"\\n"}{"点击软键盘回车时的回调\\n"}{"\\n"}{"### \`bindinput\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputInputEvent {\\n  value: string;\\n  selectionStart: number;\\n  selectionEnd: number;\\n  isComposing?: boolean;\\n}\\nbindinput?: (e: BaseEvent<'bindinput', InputInputEvent>) => void;\\n\`\`\`\\n"}{"\\n"}{"输入框内容变化的回调\\n"}{"\\n"}{"### \`bindselection\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputSelectionEvent {\\n  selectionStart: number;\\n  selectionEnd: number;\\n}\\nbindselection?: (e: BaseEvent<'bindselection', InputSelectionEvent>) => void;\\n\`\`\`\\n"}{"\\n"}{"输入框光标变化的回调\\n"}{"\\n"}{"## 方法\\n"}{"\\n"}{"前端可以通过 [SelectorQuery](/api/lynx-api/nodes-ref/nodes-ref-invoke.html) API 执行元件的方法。\\n"}{"\\n"}{"### \`focus\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputFocusMethod extends BaseMethod {\\n  method: 'focus';\\n}\\n\`\`\`\\n"}{"\\n"}{"控制输入框主动聚焦\\n"}{"\\n"}{"### \`blur\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputBlurMethod extends BaseMethod {\\n  method: 'blur';\\n}\\n\`\`\`\\n"}{"\\n"}{"控制输入框主动取消聚焦\\n"}{"\\n"}{"### \`getValue\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputGetValueMethod extends BaseMethod {\\n  method: 'getValue';\\n  success?: Callback<{\\n    value: string;\\n    selectionStart: number;\\n    selectionEnd: number;\\n    isComposing: boolean;\\n  }>;\\n}\\n\`\`\`\\n"}{"\\n"}{"获取输入框的内容\\n"}{"\\n"}{"### \`setValue\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputSetValueMethod extends BaseMethod {\\n  method: 'setValue';\\n  params: {\\n    value: string;\\n  };\\n}\\n\`\`\`\\n"}{"\\n"}{"主动设置输入框的内容\\n"}{"\\n"}{"### \`setSelectionRange\`\\n"}{"\\n"}{"\`\`\`ts\\nexport interface InputSetSelectionRangeMethod extends BaseMethod {\\n  method: 'setSelectionRange';\\n  params: {\\n    selectionStart: number;\\n    selectionEnd: number;\\n  };\\n}\\n\`\`\`\\n"}{"\\n"}{"主动设置输入框的光标\\n"}{"\\n"}{"## 兼容性\\n"}{"\\n"}{"<LegacyCompatTable metadata=\\"elements/input\\" />\\n"}</>;
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
"
`;
